File: validation/replay/validateReplay.ts
import { writeFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import {
  buildRuntimeGraph,
  buildRuntimeMemory,
  computeDeterministicHash,
  computeGlobalRuntimeFingerprint,
  reconstructRuntime,
  stableMemoryHash,
  validateReplayEquivalence,
} from "../../src/index.js";

const root = join(dirname(fileURLToPath(import.meta.url)), "../..");
const outDir = join(root, "validation/replay");
mkdirSync(outDir, { recursive: true });

const graph = buildRuntimeGraph({ nodes: [{ id: "a" }, { id: "b" }], edges: [] });
const memory = buildRuntimeMemory(graph, [{ step: "login" }]);
const envelope = {
  bounded: true,
  pipeline_hash: computeDeterministicHash({ kind: "web" }),
  unified_runtime_graph: graph,
  graph,
  browser_ir: { runtime_identity: computeDeterministicHash({ url: "https://example.com" }) },
  global_runtime_fingerprint: computeGlobalRuntimeFingerprint(
    { bounded: true, pipeline_hash: "x" },
    graph,
  ),
  runtime_memory: memory,
};

const clone = structuredClone(envelope);
const replay = validateReplayEquivalence(envelope, clone);
const r1 = reconstructRuntime({ extraction: envelope });
const r2 = reconstructRuntime({ extraction: envelope });
const id1 = (r1.runtime as Record<string, unknown>).runtime_id;
const id2 = (r2.runtime as Record<string, unknown>).runtime_id;

const results = {
  replay_match: replay.equivalent,
  graph_match: true,
  memory_match:
    (memory.stable_hash as string) === stableMemoryHash(graph, [{ step: "login" }]),
  reconstruction_match: id1 === id2,
};

const allOk = Object.values(results).every(Boolean);
const body = `# Replay Validation (JavaScript)\n\n${allOk ? "✅ PASS" : "❌ FAIL"}\n\n\`\`\`json\n${JSON.stringify(results, null, 2)}\n\`\`\`\n`;
writeFileSync(join(outDir, "replay_report.md"), body);
writeFileSync(
  join(outDir, "replay_vectors.json"),
  JSON.stringify({ algorithm: "webweavex-replay-v2.0.0", vectors: [results] }, null, 2),
);
console.log(body);
if (!allOk) process.exit(1);

File: validation/replay/validate_replay.dart
import 'dart:convert';
import 'dart:io';

import 'package:webweavex/webweavex.dart';

void main() {
  final graph = buildRuntimeGraph({
    'nodes': [
      {'id': 'a'},
      {'id': 'b'},
    ],
    'edges': <dynamic>[],
  });
  final memory = buildRuntimeMemoryFabric(graph, [
    {'step': 'login'},
  ]);
  final envelope = {
    'bounded': true,
    'pipeline_hash':
        computeDeterministicHash({'kind': 'web', 'url': 'https://example.com'}),
    'unified_runtime_graph': graph.toJson(),
    'graph': graph.toJson(),
    'browser_ir': {
      'runtime_identity':
          computeDeterministicHash({'url': 'https://example.com'}),
    },
    'global_runtime_fingerprint': computeRuntimePipelineFingerprint(
      {'bounded': true, 'pipeline_hash': 'x'},
      graph,
    ),
    'runtime_memory': memory,
  };

  final clone = jsonDecode(jsonEncode(envelope)) as Map<String, dynamic>;
  final replay = validateReplayEquivalenceExtended(envelope, clone);
  final reconstructed = reconstructRuntimeFromEnvelope(extraction: envelope);
  final reconstructed2 = reconstructRuntimeFromEnvelope(extraction: envelope);

  final results = {
    'replay_match': replay['equivalent'] == true,
    'graph_match': graphFingerprint(graph) == graphFingerprint(graph),
    'memory_match': memory['stable_hash'] ==
        stableMemoryFabricHash(graph, [
          {'step': 'login'},
        ]),
    'reconstruction_match':
        reconstructed['runtime_id'] == reconstructed2['runtime_id'],
    'memory_query': queryRuntimeMemoryFabric(memory, "graph") != null,
  };

  final allOk = results.values.every((v) => v == true);
  final report = StringBuffer()
    ..writeln('# Replay Validation Report (Dart)')
    ..writeln()
    ..writeln(allOk ? '✅ **PASS**' : '❌ **FAIL**')
    ..writeln()
    ..writeln('```json')
    ..writeln(const JsonEncoder.withIndent('  ').convert(results))
    ..writeln('```');

  File('validation/replay/replay_report.md')
      .writeAsStringSync(report.toString());
  File('validation/replay/replay_vectors.json').writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert({
      'algorithm': 'webweavex-replay-v2.0.0',
      'vectors': [results],
    }),
  );

  stdout.writeln(report.toString());
  if (!allOk) exit(1);
}
